addons/mail/controllers/discuss/gif.py

```diff
@@ -0,0 +1,58 @@
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+import json
+import requests
+import urllib.parse
+from odoo.http import request, route, Controller
+
+
+class DiscussGifController(Controller):
+
+    def _request_gifs(self, url):
+        try:
+            response = requests.get(url, timeout=3)
+        except requests.exceptions.RequestException:
+            return False
+        if response.status_code != requests.codes.ok:
+            return False
+        return response
+
+    @route("/discuss/gif/search", type="json", auth="public", cors="*")
+    def search(self, search_term, locale="en", country="US", position=False):
+        config = request.env["ir.config_parameter"].sudo().tenor_query_params()
+        url = f"https://tenor.googleapis.com/v2/search?q={urllib.parse.quote_plus(search_term)}&key={config.get('api_key')}&client_key={config.get('client_key')}&limit={config.get('tenor_gif_limit')}&contentfilter={config.get('content_filter')}&locale={urllib.parse.quote_plus(locale)}&country={urllib.parse.quote_plus(country)}&media_filter=tinygif"
+        if position:
+            url = f"{url}&pos={position}"
+        response = self._request_gifs(url)
+        if response:
+            return json.loads(response.content)
+
+    @route("/discuss/gif/categories", type="json", auth="public", cors="*")
+    def categories(self, locale="en", country="US"):
+        config = request.env["ir.config_parameter"].sudo().tenor_query_params()
+        url = f"https://tenor.googleapis.com/v2/categories?key={config.get('api_key')}&client_key={config.get('client_key')}&contentfilter={config.get('content_filter')}&locale={urllib.parse.quote_plus(locale)}&country={urllib.parse.quote_plus(country)}"
+        response = self._request_gifs(url)
+        if response:
+            return json.loads(response.content)
+
+    @route("/discuss/gif/add_favorite", type="json", auth="user")
+    def add_favorite(self, tenor_gif_id):
+        request.env["discuss.gif.favorite"].create({"tenor_gif_id": tenor_gif_id})
```

Comment:
shouldn't try to catch the unique exception here?

Reply:
I do not understand what you mean here. Could you elaborate?

Reply:
what happens if I call this route twice with the same gif id?

Reply:
It should be blocked by the sql constraints? That should display the error "User should not have duplicated favorite GIF".
Let's say I magically guess what exception is raised, what should I do with it?